[javascript] Error - no Node adapter class registered for XmlPropRef

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

**Affects PMD Version:** 
Version 6.42.0-SNAPSHOT

Make sure, to test with the latest PMD version.

**Description:**
I get the error below whenever I try to run the custom rule below on the Javascript file (Content included below) 

**Exception Stacktrace:**

```
Dec 27, 2021 11:53:04 PM net.sourceforge.pmd.processor.PmdRunnable addError
FINE: Error while processing file: /usr/local/google/home/martinsu/work/shg-eng-excellence/apexrules/src/test/java/com/google/rules/pmd/javascript/StandardJSTemplate.js
net.sourceforge.pmd.PMDException: Error while processing /usr/local/google/home/martinsu/work/shg-eng-excellence/apexrules/src/test/java/com/google/rules/pmd/javascript/StandardJSTemplate.js
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:128)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:100)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:62)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:85)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:830)
Caused by: java.lang.IllegalArgumentException: There is no Node adapter class registered for the Node class: class org.mozilla.javascript.ast.XmlPropRef
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptTreeBuilder.createNodeAdapter(EcmascriptTreeBuilder.java:166)
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptTreeBuilder.buildInternal(EcmascriptTreeBuilder.java:191)
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptTreeBuilder.visit(EcmascriptTreeBuilder.java:217)
	at org.mozilla.javascript.ast.XmlPropRef.visit(XmlPropRef.java:84)
	at org.mozilla.javascript.ast.ExpressionStatement.visit(ExpressionStatement.java:124)
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptTreeBuilder.buildInternal(EcmascriptTreeBuilder.java:205)
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptTreeBuilder.visit(EcmascriptTreeBuilder.java:217)
	at org.mozilla.javascript.ast.ExpressionStatement.visit(ExpressionStatement.java:123)
	at org.mozilla.javascript.ast.ScriptNode.visit(ScriptNode.java:322)
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptTreeBuilder.buildInternal(EcmascriptTreeBuilder.java:205)
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptTreeBuilder.build(EcmascriptTreeBuilder.java:177)
	at net.sourceforge.pmd.lang.ecmascript.ast.EcmascriptParser.parse(EcmascriptParser.java:70)
	at net.sourceforge.pmd.lang.ecmascript.Ecmascript3Parser.parse(Ecmascript3Parser.java:47)
	at net.sourceforge.pmd.lang.AbstractParser.doParse(AbstractParser.java:44)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:136)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:200)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:118)
	... 10 more
```

**Code Sample demonstrating the issue:**

```js
import { LightningElement, api, wire, track } from 'lwc';
import { ShowToastEvent } from 'lightning/platformShowToastEvent';

export default class InteractionScore extends LightningElement {
  @api recordId;
  @track
  relatedRecords;
  @track showMessage;
  /**
   * @description This is a wire adapter used to get case Object Info
   * @param
   */
  @wire(getRelatedRecordByCaseId, {
    caseId: '$recordId'
  })
  caseObjectInfo({
    error,
    data
  }) {
    if (data) {

  }
  handleDetailAnalysisClick(event) {
  }
  /**
   * @description Method to show toast incase of any error.
   */
  showErrorToast(errorMessage) {
    const evt = new ShowToastEvent({
      title: 'Error Occurred',
      message: errorMessage,
      variant: 'error'
    });
    this.dispatchEvent(evt);
  }
}
}
```

**Steps to reproduce:**
Any Javascript rule can be used to reproduce this issue as it fails for any rule on this test file, StandardJSTemplate.js, so please feel free to run this file using any Javascript rule. You will reproduce the issue.

Please provide detailed steps for how we can reproduce the bug.

1.  ... (e.g. if you're using maven: `mvn clean verify`) Ran mvn clean install to generate the jar that was copied over to PMD/bin
2.  ... Then tested the specific custom rule with this command: ./test.sh -x JavascriptMethodLength.xml -t StandardJSTemplate.js
3. test.sh is a shell script that runs the pmd command for the specific rule contained in the: JavascriptMethodLength.xml on the Javascript test file: StandardJSTemplate.js
4. Note that this rule works for other Javascript files that do not contain annotations like @wire, @api, @track. It gives the error above for files with these annotations.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Running PMD through CLI using the shell script test.sh

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -14,6 +14,13 @@ This is a {{ site.pmd.release_type }} release.
 
 ### New and noteworthy
 
+#### Javascript: Rhino updated to latest version 1.7.14
+
+[Rhino](https://github.com/mozilla/rhino), the implementation of JavaScript we use
+for parsing JavaScript code, has been updated to the latest version 1.7.14.
+Now language features like template strings can be parsed. However Rhino does
+not support all features of the latest EcmaScript standard.
+
 #### Modified rules
 
 *   The Apex rule {% rule "apex/documentation/ApexDoc" %} has a new property `reportProperty`.

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -29,6 +29,8 @@ This is a {{ site.pmd.release_type }} release.
     *   [#3686](https://github.com/pmd/pmd/issues/3686): \[java] ReturnEmptyCollectionRatherThanNull - false negative with conditioned returns
 *   java-performance
     *   [#3492](https://github.com/pmd/pmd/issues/3492): \[java] UselessStringValueOf: False positive when there is no initial String to append to
+*   javascript
+    *   [#3703](https://github.com/pmd/pmd/issues/3703): \[javascript] Error - no Node adapter class registered for XmlPropRef
 
 ### API Changes
 

--- a/pmd-javascript/pom.xml
+++ b/pmd-javascript/pom.xml
@@ -82,7 +82,7 @@
         <dependency>
             <groupId>org.mozilla</groupId>
             <artifactId>rhino</artifactId>
-            <version>1.7.13</version>
+            <version>1.7.14</version>
         </dependency>
         <dependency>
             <groupId>commons-io</groupId>

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTBigIntLiteral.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTBigIntLiteral.java
@@ -1,0 +1,26 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import java.math.BigInteger;
+
+import org.mozilla.javascript.ast.BigIntLiteral;
+
+public final class ASTBigIntLiteral extends AbstractEcmascriptNode<BigIntLiteral> {
+
+    ASTBigIntLiteral(BigIntLiteral bigIntLiteral) {
+        super(bigIntLiteral);
+        super.setImage(bigIntLiteral.getValue());
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+
+    public BigInteger getNumber() {
+        return node.getBigInt();
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTErrorNode.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTErrorNode.java
@@ -1,0 +1,24 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.ErrorNode;
+
+public final class ASTErrorNode extends AbstractEcmascriptNode<ErrorNode> {
+
+    ASTErrorNode(ErrorNode errorNode) {
+        super(errorNode);
+        super.setImage(errorNode.getMessage());
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+
+    public String getMessage() {
+        return node.getMessage();
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTGeneratorExpression.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTGeneratorExpression.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.GeneratorExpression;
+
+public final class ASTGeneratorExpression extends AbstractEcmascriptNode<GeneratorExpression> {
+
+    ASTGeneratorExpression(GeneratorExpression generatorExpression) {
+        super(generatorExpression);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTGeneratorExpressionLoop.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTGeneratorExpressionLoop.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.GeneratorExpressionLoop;
+
+public final class ASTGeneratorExpressionLoop extends AbstractEcmascriptNode<GeneratorExpressionLoop> {
+
+    ASTGeneratorExpressionLoop(GeneratorExpressionLoop generatorExpressionLoop) {
+        super(generatorExpressionLoop);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTScriptNode.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTScriptNode.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.ScriptNode;
+
+public final class ASTScriptNode extends AbstractEcmascriptNode<ScriptNode> {
+
+    ASTScriptNode(ScriptNode scriptNode) {
+        super(scriptNode);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTTaggedTemplateLiteral.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTTaggedTemplateLiteral.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.TaggedTemplateLiteral;
+
+public final class ASTTaggedTemplateLiteral extends AbstractEcmascriptNode<TaggedTemplateLiteral> {
+
+    ASTTaggedTemplateLiteral(TaggedTemplateLiteral taggedTemplateLiteral) {
+        super(taggedTemplateLiteral);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTTemplateCharacters.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTTemplateCharacters.java
@@ -1,0 +1,23 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.TemplateCharacters;
+
+public final class ASTTemplateCharacters extends AbstractEcmascriptNode<TemplateCharacters> {
+
+    ASTTemplateCharacters(TemplateCharacters templateCharacters) {
+        super(templateCharacters);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+
+    public String getValue() {
+        return node.getValue();
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTTemplateLiteral.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTTemplateLiteral.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.TemplateLiteral;
+
+public final class ASTTemplateLiteral extends AbstractEcmascriptNode<TemplateLiteral> {
+
+    ASTTemplateLiteral(TemplateLiteral templateLiteral) {
+        super(templateLiteral);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTUnaryExpression.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTUnaryExpression.java
@@ -15,11 +15,7 @@ public class ASTUnaryExpression extends AbstractEcmascriptNode<UnaryExpression> 
     @InternalApi
     public ASTUnaryExpression(UnaryExpression unaryExpression) {
         super(unaryExpression);
-        if (unaryExpression.getOperator() == Token.VOID) {
-            super.setImage("void");
-        } else {
-            super.setImage(AstRoot.operatorToString(unaryExpression.getOperator()));
-        }
+        super.setImage(AstRoot.operatorToString(unaryExpression.getOperator()));
     }
 
     @Override
@@ -32,10 +28,10 @@ public class ASTUnaryExpression extends AbstractEcmascriptNode<UnaryExpression> 
     }
 
     public boolean isPrefix() {
-        return node.isPrefix();
+        return !isPostfix();
     }
 
     public boolean isPostfix() {
-        return node.isPostfix();
+        return node.getOperator() == Token.INC || node.getOperator() == Token.DEC;
     }
 }

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTUpdateExpression.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTUpdateExpression.java
@@ -1,0 +1,34 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.AstNode;
+import org.mozilla.javascript.ast.AstRoot;
+import org.mozilla.javascript.ast.UpdateExpression;
+
+public final class ASTUpdateExpression extends AbstractEcmascriptNode<UpdateExpression> {
+
+    ASTUpdateExpression(UpdateExpression updateExpression) {
+        super(updateExpression);
+        super.setImage(AstRoot.operatorToString(updateExpression.getOperator()));
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+
+    public String getOperator() {
+        return AstNode.operatorToString(node.getOperator());
+    }
+
+    public boolean isPostfix() {
+        return node.isPostfix();
+    }
+
+    public boolean isPrefix() {
+        return node.isPrefix();
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlElemRef.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlElemRef.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.XmlElemRef;
+
+public final class ASTXmlElemRef extends AbstractEcmascriptNode<XmlElemRef> {
+
+    ASTXmlElemRef(XmlElemRef xmlElemRef) {
+        super(xmlElemRef);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlLiteral.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlLiteral.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.XmlLiteral;
+
+public final class ASTXmlLiteral extends AbstractEcmascriptNode<XmlLiteral> {
+
+    ASTXmlLiteral(XmlLiteral xmlLiteral) {
+        super(xmlLiteral);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlPropRef.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlPropRef.java
@@ -7,12 +7,9 @@ package net.sourceforge.pmd.lang.ecmascript.ast;
 import org.mozilla.javascript.ast.Name;
 import org.mozilla.javascript.ast.XmlPropRef;
 
-import net.sourceforge.pmd.annotation.InternalApi;
+public final class ASTXmlPropRef extends AbstractEcmascriptNode<XmlPropRef> {
 
-public class ASTXmlPropRef extends AbstractEcmascriptNode<XmlPropRef> {
-    @Deprecated
-    @InternalApi
-    public ASTXmlPropRef(XmlPropRef xmlPropRef) {
+    ASTXmlPropRef(XmlPropRef xmlPropRef) {
         super(xmlPropRef);
 
         Name propName = xmlPropRef.getPropName();

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlPropRef.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTXmlPropRef.java
@@ -1,0 +1,28 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.Name;
+import org.mozilla.javascript.ast.XmlPropRef;
+
+import net.sourceforge.pmd.annotation.InternalApi;
+
+public class ASTXmlPropRef extends AbstractEcmascriptNode<XmlPropRef> {
+    @Deprecated
+    @InternalApi
+    public ASTXmlPropRef(XmlPropRef xmlPropRef) {
+        super(xmlPropRef);
+
+        Name propName = xmlPropRef.getPropName();
+        if (propName != null) {
+            super.setImage(propName.getIdentifier());
+        }
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTYield.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/ASTYield.java
@@ -1,0 +1,19 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ecmascript.ast;
+
+import org.mozilla.javascript.ast.Yield;
+
+public final class ASTYield extends AbstractEcmascriptNode<Yield> {
+
+    ASTYield(Yield yield) {
+        super(yield);
+    }
+
+    @Override
+    public Object jjtAccept(EcmascriptParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+}

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/EcmascriptTreeBuilder.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/EcmascriptTreeBuilder.java
@@ -18,6 +18,7 @@ import org.mozilla.javascript.ast.ArrayLiteral;
 import org.mozilla.javascript.ast.Assignment;
 import org.mozilla.javascript.ast.AstNode;
 import org.mozilla.javascript.ast.AstRoot;
+import org.mozilla.javascript.ast.BigIntLiteral;
 import org.mozilla.javascript.ast.Block;
 import org.mozilla.javascript.ast.BreakStatement;
 import org.mozilla.javascript.ast.CatchClause;
@@ -28,11 +29,14 @@ import org.mozilla.javascript.ast.DoLoop;
 import org.mozilla.javascript.ast.ElementGet;
 import org.mozilla.javascript.ast.EmptyExpression;
 import org.mozilla.javascript.ast.EmptyStatement;
+import org.mozilla.javascript.ast.ErrorNode;
 import org.mozilla.javascript.ast.ExpressionStatement;
 import org.mozilla.javascript.ast.ForInLoop;
 import org.mozilla.javascript.ast.ForLoop;
 import org.mozilla.javascript.ast.FunctionCall;
 import org.mozilla.javascript.ast.FunctionNode;
+import org.mozilla.javascript.ast.GeneratorExpression;
+import org.mozilla.javascript.ast.GeneratorExpressionLoop;
 import org.mozilla.javascript.ast.IfStatement;
 import org.mozilla.javascript.ast.InfixExpression;
 import org.mozilla.javascript.ast.KeywordLiteral;
@@ -51,21 +55,29 @@ import org.mozilla.javascript.ast.PropertyGet;
 import org.mozilla.javascript.ast.RegExpLiteral;
 import org.mozilla.javascript.ast.ReturnStatement;
 import org.mozilla.javascript.ast.Scope;
+import org.mozilla.javascript.ast.ScriptNode;
 import org.mozilla.javascript.ast.StringLiteral;
 import org.mozilla.javascript.ast.SwitchCase;
 import org.mozilla.javascript.ast.SwitchStatement;
+import org.mozilla.javascript.ast.TaggedTemplateLiteral;
+import org.mozilla.javascript.ast.TemplateCharacters;
+import org.mozilla.javascript.ast.TemplateLiteral;
 import org.mozilla.javascript.ast.ThrowStatement;
 import org.mozilla.javascript.ast.TryStatement;
 import org.mozilla.javascript.ast.UnaryExpression;
+import org.mozilla.javascript.ast.UpdateExpression;
 import org.mozilla.javascript.ast.VariableDeclaration;
 import org.mozilla.javascript.ast.VariableInitializer;
 import org.mozilla.javascript.ast.WhileLoop;
 import org.mozilla.javascript.ast.WithStatement;
 import org.mozilla.javascript.ast.XmlDotQuery;
+import org.mozilla.javascript.ast.XmlElemRef;
 import org.mozilla.javascript.ast.XmlExpression;
+import org.mozilla.javascript.ast.XmlLiteral;
 import org.mozilla.javascript.ast.XmlMemberGet;
 import org.mozilla.javascript.ast.XmlPropRef;
 import org.mozilla.javascript.ast.XmlString;
+import org.mozilla.javascript.ast.Yield;
 
 import net.sourceforge.pmd.annotation.InternalApi;
 import net.sourceforge.pmd.lang.ast.Node;
@@ -83,6 +95,7 @@ public final class EcmascriptTreeBuilder implements NodeVisitor {
         register(ArrayLiteral.class, ASTArrayLiteral.class);
         register(Assignment.class, ASTAssignment.class);
         register(AstRoot.class, ASTAstRoot.class);
+        register(BigIntLiteral.class, ASTBigIntLiteral.class);
         register(Block.class, ASTBlock.class);
         register(BreakStatement.class, ASTBreakStatement.class);
         register(CatchClause.class, ASTCatchClause.class);
@@ -93,13 +106,17 @@ public final class EcmascriptTreeBuilder implements NodeVisitor {
         register(ElementGet.class, ASTElementGet.class);
         register(EmptyExpression.class, ASTEmptyExpression.class);
         register(EmptyStatement.class, ASTEmptyStatement.class);
+        register(ErrorNode.class, ASTErrorNode.class);
         register(ExpressionStatement.class, ASTExpressionStatement.class);
         register(ForInLoop.class, ASTForInLoop.class);
         register(ForLoop.class, ASTForLoop.class);
         register(FunctionCall.class, ASTFunctionCall.class);
         register(FunctionNode.class, ASTFunctionNode.class);
+        register(GeneratorExpression.class, ASTGeneratorExpression.class);
+        register(GeneratorExpressionLoop.class, ASTGeneratorExpressionLoop.class);
         register(IfStatement.class, ASTIfStatement.class);
         register(InfixExpression.class, ASTInfixExpression.class);
+        // - not a real node - register(Jump.class, ASTJump.class);
         register(KeywordLiteral.class, ASTKeywordLiteral.class);
         register(Label.class, ASTLabel.class);
         register(LabeledStatement.class, ASTLabeledStatement.class);
@@ -114,21 +131,29 @@ public final class EcmascriptTreeBuilder implements NodeVisitor {
         register(RegExpLiteral.class, ASTRegExpLiteral.class);
         register(ReturnStatement.class, ASTReturnStatement.class);
         register(Scope.class, ASTScope.class);
+        register(ScriptNode.class, ASTScriptNode.class);
         register(StringLiteral.class, ASTStringLiteral.class);
         register(SwitchCase.class, ASTSwitchCase.class);
         register(SwitchStatement.class, ASTSwitchStatement.class);
+        register(TaggedTemplateLiteral.class, ASTTaggedTemplateLiteral.class);
+        register(TemplateCharacters.class, ASTTemplateCharacters.class);
+        register(TemplateLiteral.class, ASTTemplateLiteral.class);
         register(ThrowStatement.class, ASTThrowStatement.class);
         register(TryStatement.class, ASTTryStatement.class);
         register(UnaryExpression.class, ASTUnaryExpression.class);
+        register(UpdateExpression.class, ASTUpdateExpression.class);
         register(VariableDeclaration.class, ASTVariableDeclaration.class);
         register(VariableInitializer.class, ASTVariableInitializer.class);
         register(WhileLoop.class, ASTWhileLoop.class);
         register(WithStatement.class, ASTWithStatement.class);
         register(XmlDotQuery.class, ASTXmlDotQuery.class);
+        register(XmlElemRef.class, ASTXmlElemRef.class);
         register(XmlExpression.class, ASTXmlExpression.class);
         register(XmlMemberGet.class, ASTXmlMemberGet.class);
         register(XmlPropRef.class, ASTXmlPropRef.class);
         register(XmlString.class, ASTXmlString.class);
+        register(XmlLiteral.class, ASTXmlLiteral.class);
+        register(Yield.class, ASTYield.class);
     }
 
     private List<ParseProblem> parseProblems;
@@ -150,7 +175,7 @@ public final class EcmascriptTreeBuilder implements NodeVisitor {
     private static <T extends AstNode> void register(Class<T> nodeType,
             Class<? extends AbstractEcmascriptNode<T>> nodeAdapterType) {
         try {
-            NODE_TYPE_TO_NODE_ADAPTER_TYPE.put(nodeType, nodeAdapterType.getConstructor(nodeType));
+            NODE_TYPE_TO_NODE_ADAPTER_TYPE.put(nodeType, nodeAdapterType.getDeclaredConstructor(nodeType));
         } catch (SecurityException | NoSuchMethodException e) {
             throw new RuntimeException(e);
         }

--- a/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/EcmascriptTreeBuilder.java
+++ b/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/ast/EcmascriptTreeBuilder.java
@@ -64,6 +64,7 @@ import org.mozilla.javascript.ast.WithStatement;
 import org.mozilla.javascript.ast.XmlDotQuery;
 import org.mozilla.javascript.ast.XmlExpression;
 import org.mozilla.javascript.ast.XmlMemberGet;
+import org.mozilla.javascript.ast.XmlPropRef;
 import org.mozilla.javascript.ast.XmlString;
 
 import net.sourceforge.pmd.annotation.InternalApi;
@@ -126,6 +127,7 @@ public final class EcmascriptTreeBuilder implements NodeVisitor {
         register(XmlDotQuery.class, ASTXmlDotQuery.class);
         register(XmlExpression.class, ASTXmlExpression.class);
         register(XmlMemberGet.class, ASTXmlMemberGet.class);
+        register(XmlPropRef.class, ASTXmlPropRef.class);
         register(XmlString.class, ASTXmlString.class);
     }
 

--- a/pmd-javascript/src/main/resources/category/ecmascript/codestyle.xml
+++ b/pmd-javascript/src/main/resources/category/ecmascript/codestyle.xml
@@ -26,17 +26,17 @@ indicative of the bug where the assignment operator '=' was used instead of the 
             <property name="xpath">
                 <value>
 <![CDATA[
-    //IfStatement[$allowIf = false()]/child::node()[1]/descendant-or-self::node()[self::Assignment or self::UnaryExpression[$allowIncrementDecrement = false() and (@Image = "--" or @Image = "++")]]
+    //IfStatement[$allowIf = false()]/child::node()[1]/descendant-or-self::node()[self::Assignment or self::UpdateExpression[$allowIncrementDecrement = false() and @Operator = ('--', '++')]]
 |
-    //WhileLoop[$allowWhile = false()]/child::node()[1]/descendant-or-self::node()[self::Assignment or self::UnaryExpression[$allowIncrementDecrement = false() and (@Image = "--" or @Image = "++")]]
+    //WhileLoop[$allowWhile = false()]/child::node()[1]/descendant-or-self::node()[self::Assignment or self::UpdateExpression[$allowIncrementDecrement = false() and @Operator = ('--', '++')]]
 |
-    //DoLoop[$allowWhile = false()]/child::node()[2]/descendant-or-self::node()[self::Assignment or self::UnaryExpression[$allowIncrementDecrement = false() and (@Image = "--" or @Image = "++")]]
+    //DoLoop[$allowWhile = false()]/child::node()[2]/descendant-or-self::node()[self::Assignment or self::UpdateExpression[$allowIncrementDecrement = false() and @Operator = ('--', '++')]]
 |
-    //ForLoop[$allowFor = false()]/child::node()[2]/descendant-or-self::node()[self::Assignment or self::UnaryExpression[$allowIncrementDecrement = false() and (@Image = "--" or @Image = "++")]]
+    //ForLoop[$allowFor = false()]/child::node()[2]/descendant-or-self::node()[self::Assignment or self::UpdateExpression[$allowIncrementDecrement = false() and @Operator = ('--', '++')]]
 |
-   //ConditionalExpression[$allowTernary = false()]/child::node()[1]/descendant-or-self::node()[self::Assignment or self::UnaryExpression[$allowIncrementDecrement = false() and (@Image = "--" or @Image = "++")]]
+   //ConditionalExpression[$allowTernary = false()]/child::node()[1]/descendant-or-self::node()[self::Assignment or self::UpdateExpression[$allowIncrementDecrement = false() and @Operator = ('--', '++')]]
 |
-   //ConditionalExpression[$allowTernaryResults = false()]/child::node()[position() = 2 or position() = 3]/descendant-or-self::node()[self::Assignment or self::UnaryExpression[$allowIncrementDecrement = false() and (@Image = "--" or @Image = "++")]]
+   //ConditionalExpression[$allowTernaryResults = false()]/child::node()[position() = 2 or position() = 3]/descendant-or-self::node()[self::Assignment or self::UpdateExpression[$allowIncrementDecrement = false() and @Operator = ('--', '++')]]
 ]]>
                 </value>
             </property>

--- a/pmd-javascript/src/test/java/net/sourceforge/pmd/lang/ecmascript/ast/JsTreeDumpTest.java
+++ b/pmd-javascript/src/test/java/net/sourceforge/pmd/lang/ecmascript/ast/JsTreeDumpTest.java
@@ -35,4 +35,9 @@ public class JsTreeDumpTest extends BaseTreeDumpTest {
     public void decorators() {
         doTest("decorators");
     }
+
+    @Test
+    public void templateStrings() {
+        doTest("templateStrings");
+    }
 }

--- a/pmd-javascript/src/test/java/net/sourceforge/pmd/lang/ecmascript/ast/JsTreeDumpTest.java
+++ b/pmd-javascript/src/test/java/net/sourceforge/pmd/lang/ecmascript/ast/JsTreeDumpTest.java
@@ -30,4 +30,9 @@ public class JsTreeDumpTest extends BaseTreeDumpTest {
     public void jquerySelector() {
         doTest("jquery-selector");
     }
+
+    @Test
+    public void decorators() {
+        doTest("decorators");
+    }
 }

--- a/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/decorators.js
+++ b/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/decorators.js
@@ -1,0 +1,37 @@
+import { LightningElement, api, wire, track } from 'lwc';
+import { ShowToastEvent } from 'lightning/platformShowToastEvent';
+
+export default class InteractionScore extends LightningElement {
+  @api recordId;
+  @track
+  relatedRecords;
+  @track showMessage;
+  /**
+   * @description This is a wire adapter used to get case Object Info
+   * @param
+   */
+  @wire(getRelatedRecordByCaseId, {
+    caseId: '$recordId'
+  })
+  caseObjectInfo({
+    error,
+    data
+  }) {
+    if (data) {
+
+    }
+    handleDetailAnalysisClick(event) {
+    }
+    /**
+     * @description Method to show toast incase of any error.
+     */
+    showErrorToast(errorMessage) {
+      const evt = new ShowToastEvent({
+        title: 'Error Occurred',
+        message: errorMessage,
+        variant: 'error'
+      });
+      this.dispatchEvent(evt);
+    }
+  }
+}

--- a/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/decorators.txt
+++ b/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/decorators.txt
@@ -22,7 +22,6 @@
    +- EmptyStatement
    +- EmptyStatement
    +- EmptyStatement
-   +- EmptyStatement
    +- Comment
    +- EmptyStatement
    +- ExpressionStatement

--- a/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/decorators.txt
+++ b/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/decorators.txt
@@ -1,0 +1,50 @@
++- AstRoot
+   +- EmptyStatement
+   +- EmptyStatement
+   +- EmptyStatement
+   +- EmptyStatement
+   +- ExpressionStatement
+   |  +- XmlPropRef
+   |     +- Name
+   +- ExpressionStatement
+   |  +- Name
+   +- EmptyStatement
+   +- Comment
+   +- ExpressionStatement
+   |  +- FunctionCall
+   |     +- XmlPropRef
+   |     |  +- Name
+   |     +- Name
+   |     +- ObjectLiteral
+   |        +- ObjectProperty
+   |           +- Name
+   |           +- StringLiteral
+   +- EmptyStatement
+   +- EmptyStatement
+   +- EmptyStatement
+   +- EmptyStatement
+   +- Comment
+   +- EmptyStatement
+   +- ExpressionStatement
+   |  +- Assignment
+   |     +- Name
+   |     +- NewExpression
+   |        +- Name
+   |        +- ObjectLiteral
+   |           +- ObjectProperty
+   |           |  +- Name
+   |           |  +- StringLiteral
+   |           +- ObjectProperty
+   |           |  +- Name
+   |           |  +- Name
+   |           +- ObjectProperty
+   |              +- Name
+   |              +- StringLiteral
+   +- ExpressionStatement
+   |  +- FunctionCall
+   |     +- PropertyGet
+   |     |  +- KeywordLiteral
+   |     |  +- Name
+   |     +- Name
+   +- EmptyStatement
+   +- EmptyStatement

--- a/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/jquery-selector.txt
+++ b/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/jquery-selector.txt
@@ -744,7 +744,7 @@
                   |     |                 |     |        +- Name
                   |     |                 |     |        +- Name
                   |     |                 |     +- WhileLoop
-                  |     |                 |     |  +- UnaryExpression
+                  |     |                 |     |  +- UpdateExpression
                   |     |                 |     |  |  +- Name
                   |     |                 |     |  +- Scope
                   |     |                 |     |     +- ExpressionStatement
@@ -1075,7 +1075,7 @@
                   |                             |        +- Name
                   |                             +- Comment
                   |                             +- WhileLoop
-                  |                                +- UnaryExpression
+                  |                                +- UpdateExpression
                   |                                |  +- Name
                   |                                +- Scope
                   |                                   +- IfStatement
@@ -2239,7 +2239,7 @@
                   |           |     |                          |  |     |     +- Name
                   |           |     |                          |  |     |     +- InfixExpression
                   |           |     |                          |  |     |        +- InfixExpression
-                  |           |     |                          |  |     |        |  +- UnaryExpression
+                  |           |     |                          |  |     |        |  +- UpdateExpression
                   |           |     |                          |  |     |        |  |  +- Name
                   |           |     |                          |  |     |        |  +- InfixExpression
                   |           |     |                          |  |     |        |     +- Name
@@ -2267,7 +2267,7 @@
                   |           |     |                          |  |           |  |  |  +- Name
                   |           |     |                          |  |           |  |  +- NumberLiteral
                   |           |     |                          |  |           |  +- InfixExpression
-                  |           |     |                          |  |           |     +- UnaryExpression
+                  |           |     |                          |  |           |     +- UpdateExpression
                   |           |     |                          |  |           |     |  +- Name
                   |           |     |                          |  |           |     +- InfixExpression
                   |           |     |                          |  |           |        +- Name
@@ -2337,7 +2337,7 @@
                   |           |     |                          |              |     +- Name
                   |           |     |                          |              |     +- InfixExpression
                   |           |     |                          |              |        +- InfixExpression
-                  |           |     |                          |              |        |  +- UnaryExpression
+                  |           |     |                          |              |        |  +- UpdateExpression
                   |           |     |                          |              |        |  |  +- Name
                   |           |     |                          |              |        |  +- InfixExpression
                   |           |     |                          |              |        |     +- Name
@@ -2370,7 +2370,7 @@
                   |           |     |                          |                    |  |        |  +- Name
                   |           |     |                          |                    |  |        |  +- Name
                   |           |     |                          |                    |  |        +- NumberLiteral
-                  |           |     |                          |                    |  +- UnaryExpression
+                  |           |     |                          |                    |  +- UpdateExpression
                   |           |     |                          |                    |     +- Name
                   |           |     |                          |                    +- Scope
                   |           |     |                          |                       +- Comment
@@ -2521,7 +2521,7 @@
                   |           |              |           |        |        +- Name
                   |           |              |           |        |        +- Name
                   |           |              |           |        +- WhileLoop
-                  |           |              |           |           +- UnaryExpression
+                  |           |              |           |           +- UpdateExpression
                   |           |              |           |           |  +- Name
                   |           |              |           |           +- Scope
                   |           |              |           |              +- ExpressionStatement
@@ -2621,7 +2621,7 @@
                   |                 |                 |        |        +- Name
                   |                 |                 |        +- Comment
                   |                 |                 |        +- WhileLoop
-                  |                 |                 |           +- UnaryExpression
+                  |                 |                 |           +- UpdateExpression
                   |                 |                 |           |  +- Name
                   |                 |                 |           +- Scope
                   |                 |                 |              +- IfStatement
@@ -3221,7 +3221,7 @@
                   |                 |           +- ForLoop
                   |                 |           |  +- EmptyExpression
                   |                 |           |  +- InfixExpression
-                  |                 |           |  |  +- UnaryExpression
+                  |                 |           |  |  +- UpdateExpression
                   |                 |           |  |  |  +- Name
                   |                 |           |  |  +- NumberLiteral
                   |                 |           |  +- EmptyExpression
@@ -3257,7 +3257,7 @@
                   |                             +- ForLoop
                   |                             |  +- EmptyExpression
                   |                             |  +- InfixExpression
-                  |                             |  |  +- UnaryExpression
+                  |                             |  |  +- UpdateExpression
                   |                             |  |  |  +- Name
                   |                             |  |  +- Name
                   |                             |  +- EmptyExpression
@@ -3618,7 +3618,7 @@
                   |     |  +- InfixExpression
                   |     |  |  +- Name
                   |     |  |  +- Name
-                  |     |  +- UnaryExpression
+                  |     |  +- UpdateExpression
                   |     |  |  +- Name
                   |     |  +- Scope
                   |     |     +- ExpressionStatement
@@ -3662,7 +3662,7 @@
                   |     |  |        +- StringLiteral
                   |     |  +- VariableInitializer
                   |     |     +- Name
-                  |     |     +- UnaryExpression
+                  |     |     +- UpdateExpression
                   |     |        +- Name
                   |     +- ReturnStatement
                   |        +- ConditionalExpression
@@ -3869,7 +3869,7 @@
                   |           |     |        +- Name
                   |           |     |        +- Name
                   |           |     +- WhileLoop
-                  |           |     |  +- UnaryExpression
+                  |           |     |  +- UpdateExpression
                   |           |     |  |  +- Name
                   |           |     |  +- Scope
                   |           |     |     +- IfStatement
@@ -3909,7 +3909,7 @@
                   |     |  +- InfixExpression
                   |     |  |  +- Name
                   |     |  |  +- Name
-                  |     |  +- UnaryExpression
+                  |     |  +- UpdateExpression
                   |     |  |  +- Name
                   |     |  +- Scope
                   |     |     +- ExpressionStatement
@@ -3954,7 +3954,7 @@
                   |     |  +- InfixExpression
                   |     |  |  +- Name
                   |     |  |  +- Name
-                  |     |  +- UnaryExpression
+                  |     |  +- UpdateExpression
                   |     |  |  +- Name
                   |     |  +- Scope
                   |     |     +- IfStatement
@@ -4153,7 +4153,7 @@
                   |                 |     |        +- Name
                   |                 |     |        +- Name
                   |                 |     +- WhileLoop
-                  |                 |        +- UnaryExpression
+                  |                 |        +- UpdateExpression
                   |                 |        |  +- Name
                   |                 |        +- Scope
                   |                 |           +- IfStatement
@@ -4203,7 +4203,7 @@
                   |                    |        |     |        +- Name
                   |                    |        |     |        +- Name
                   |                    |        |     +- WhileLoop
-                  |                    |        |     |  +- UnaryExpression
+                  |                    |        |     |  +- UpdateExpression
                   |                    |        |     |  |  +- Name
                   |                    |        |     |  +- Scope
                   |                    |        |     |     +- IfStatement
@@ -4243,7 +4243,7 @@
                   |                    |        |        +- Name
                   |                    |        |        +- Name
                   |                    |        +- WhileLoop
-                  |                    |           +- UnaryExpression
+                  |                    |           +- UpdateExpression
                   |                    |           |  +- Name
                   |                    |           +- Scope
                   |                    |              +- IfStatement
@@ -4447,7 +4447,7 @@
                   |     |  +- InfixExpression
                   |     |  |  +- Name
                   |     |  |  +- Name
-                  |     |  +- UnaryExpression
+                  |     |  +- UpdateExpression
                   |     |  |  +- Name
                   |     |  +- Scope
                   |     |     +- IfStatement
@@ -4506,14 +4506,14 @@
                   |     |           |     +- ExpressionStatement
                   |     |           |     |  +- Assignment
                   |     |           |     |     +- Name
-                  |     |           |     |     +- UnaryExpression
+                  |     |           |     |     +- UpdateExpression
                   |     |           |     |        +- Name
                   |     |           |     +- ForLoop
                   |     |           |     |  +- EmptyExpression
                   |     |           |     |  +- InfixExpression
                   |     |           |     |  |  +- Name
                   |     |           |     |  |  +- Name
-                  |     |           |     |  +- UnaryExpression
+                  |     |           |     |  +- UpdateExpression
                   |     |           |     |  |  +- Name
                   |     |           |     |  +- Scope
                   |     |           |     |     +- IfStatement
@@ -4729,7 +4729,7 @@
                   |     |           |  |  |        +- Name
                   |     |           |  |  |        +- Name
                   |     |           |  |  +- KeywordLiteral
-                  |     |           |  +- UnaryExpression
+                  |     |           |  +- UpdateExpression
                   |     |           |  |  +- Name
                   |     |           |  +- Scope
                   |     |           |     +- IfStatement
@@ -4770,7 +4770,7 @@
                   |     |           |     |     |  |     +- Name
                   |     |           |     |     |  |     +- ElementGet
                   |     |           |     |     |  |        +- Name
-                  |     |           |     |     |  |        +- UnaryExpression
+                  |     |           |     |     |  |        +- UpdateExpression
                   |     |           |     |     |  |           +- Name
                   |     |           |     |     |  +- Scope
                   |     |           |     |     |     +- IfStatement
@@ -4811,7 +4811,7 @@
                   |     |           |           |  |        +- Name
                   |     |           |           |  +- Scope
                   |     |           |           |     +- ExpressionStatement
-                  |     |           |           |        +- UnaryExpression
+                  |     |           |           |        +- UpdateExpression
                   |     |           |           |           +- Name
                   |     |           |           +- IfStatement
                   |     |           |              +- Name
@@ -4852,7 +4852,7 @@
                   |     |           |     |  |     +- Name
                   |     |           |     |  |     +- ElementGet
                   |     |           |     |  |        +- Name
-                  |     |           |     |  |        +- UnaryExpression
+                  |     |           |     |  |        +- UpdateExpression
                   |     |           |     |  |           +- Name
                   |     |           |     |  +- Scope
                   |     |           |     |     +- ExpressionStatement
@@ -4872,7 +4872,7 @@
                   |     |           |     |     |  |  +- NumberLiteral
                   |     |           |     |     |  +- Scope
                   |     |           |     |     |     +- WhileLoop
-                  |     |           |     |     |        +- UnaryExpression
+                  |     |           |     |     |        +- UpdateExpression
                   |     |           |     |     |        |  +- Name
                   |     |           |     |     |        +- Scope
                   |     |           |     |     |           +- IfStatement
@@ -5000,7 +5000,7 @@
                   |     |     |        +- Name
                   |     |     |        +- Name
                   |     |     +- WhileLoop
-                  |     |     |  +- UnaryExpression
+                  |     |     |  +- UpdateExpression
                   |     |     |  |  +- Name
                   |     |     |  +- Scope
                   |     |     |     +- ExpressionStatement
@@ -5229,7 +5229,7 @@
                   |     |     |           +- Name
                   |     |     |           +- Name
                   |     |     +- WhileLoop
-                  |     |        +- UnaryExpression
+                  |     |        +- UpdateExpression
                   |     |        |  +- Name
                   |     |        +- Scope
                   |     |           +- ExpressionStatement

--- a/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/templateStrings.js
+++ b/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/templateStrings.js
@@ -1,0 +1,4 @@
+function() {
+    let a = "World";
+    let b = `Hello $a`;
+}

--- a/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/templateStrings.txt
+++ b/pmd-javascript/src/test/resources/net/sourceforge/pmd/lang/ecmascript/ast/testdata/templateStrings.txt
@@ -1,0 +1,12 @@
++- AstRoot
+   +- FunctionNode
+      +- Block
+         +- VariableDeclaration
+         |  +- VariableInitializer
+         |     +- Name
+         |     +- StringLiteral
+         +- VariableDeclaration
+            +- VariableInitializer
+               +- Name
+               +- TemplateLiteral
+                  +- TemplateCharacters

